### PR TITLE
[cmd-parser] simplify `ParseAsHexString()` & `Cli::UdpExample::ProcessSend()`

### PR DESCRIPTION
**Pull request by abtink**
This commit enhances `CmdLineParser` methods that parse a hex string
adding a new version which allows a (longer) hex string to be parsed
in short segments sequentially. The unit test `test_cmd_line_parser`
is also updated to verify the behavior of the newly added method.

This commit also updates CLI `UdpExample` implementation to use the
new method for parsing message payload data from input hex string and
simplifies how `ProcessSend()` parses different arguments indicating
message payload options.

### CODE REVIEW

File: src/core/utils/parse_cmdline.hpp [hunk 1]
```diff
@@ -251,24 +247,45 @@ template <uint16_t kBufferSize> static otError ParseAsHexString(const char *aStr
 /**
  * This function parses a hex string into a byte array.
  *
- * If @p aMode disallows truncation (`kDisallowTruncate`), this function verifies that parses hex string bytes fit in
- * @p aBuffer with its given size in @aSize. Otherwise when @p aMode allows truncation, extra bytes after @p aSize bytes
- * are ignored.
+ * This function verifies that parses hex string bytes fit in @p aBuffer with its given @p aSize.
```
Comment (jwhui) [inline src/core/utils/parse_cmdline.hpp:250]:
```suggestion
 * This function verifies that the parsed hex string bytes fit in @p aBuffer with its given @p aSize.
```

File: src/core/utils/parse_cmdline.hpp [hunk 2]
```diff
@@ -251,24 +247,45 @@ template <uint16_t kBufferSize> static otError ParseAsHexString(const char *aStr
...
- * @retval kErrorInvalidArgs The string does not contain valid format or too many bytes (if truncation not allowed)
+ * @retval kErrorInvalidArgs The string does not contain valid format or too many bytes.
+ *
+ */
+otError ParseAsHexString(const char *aString, uint16_t &aSize, uint8_t *aBuffer);
+
+/**
+ * This function parses a segment of a hex string up to a given size.
+ *
+ * This function allows a longer hex string to be parsed and read in smaller segments into a giver buffer. If the
```
Comment (jwhui) [inline src/core/utils/parse_cmdline.hpp:269]:
```suggestion
 * This function allows a longer hex string to be parsed and read in smaller segments into a given buffer. If the
```

File: src/core/utils/parse_cmdline.hpp [hunk 3]
```diff
@@ -251,24 +247,45 @@ template <uint16_t kBufferSize> static otError ParseAsHexString(const char *aStr
...
+ * entire hex string bytes can fit in the given @p aBuffer with its @p aSize, they are copied into @p aBuffer and
+ * function returns `kErrorNone`. Otherwise, @p aSize bytes are read and copied and function returns `kErrorPending`
+ * to indicate that there are more bytes to parse. The @p aString is also updated to skip over the parsed segment.
+ *
+ * This function correctly handles hex strings with even or odd length. For example, "AABBCCDD" (with even length) is
+ * parsed as {0xaa, 0xbb, 0xcc, 0xdd} and "123" (with odd length) is parsed as {0x01, 0x23}.
+ *
+ * @param[inout] aString     A reference to string to parse. On successful parse, updated to skip parsed digits.
+ * @param[inout] aSize       On entry indicates the segment size (number of bytes on @p aBuffer).
```
Comment (jwhui) [inline src/core/utils/parse_cmdline.hpp:278]:
```suggestion
 * @param[inout] aSize       On entry indicates the segment size (number of bytes in @p aBuffer).
```

File: src/core/utils/parse_cmdline.hpp [hunk 4]
```diff
@@ -523,23 +540,19 @@ class Arg
     /**
      * This method parses the argument as a hex string into a byte array.
      *
-     * If @p aMode disallows truncation (`kDisallowTruncate`), this method verifies that parses hex string bytes fit in
-     * @p aBuffer with its given size in @aSize. Otherwise when @p aMode allows truncation, extra bytes after @p aSize
-     * bytes are ignored.
+     * This method verifies that parses hex string bytes fit in @p aBuffer with its given @p aSize.
```
Comment (jwhui) [inline src/core/utils/parse_cmdline.hpp:543]:
```suggestion
     * This method verifies that the parsed hex string bytes fit in @p aBuffer with its given @p aSize.
```

File: src/core/utils/parse_cmdline.cpp [hunk 5]
```diff
@@ -275,69 +275,86 @@ Error ParseAsIp6Prefix(const char *aString, otIp6Prefix &aPrefix)
...
     }
 
-    // Handle the case where number of chars in hex string is odd.
-    numChars = hexLength & 1;
+    if ((stringLength & 1) != 0)
+    {
+        // If number of chars in hex string is odd, we jump to `middle`
+        // label in the loop to skip parsing the first digit.
+
+        goto middle;
```
Comment (jwhui) [inline src/core/utils/parse_cmdline.cpp:311]: Can we avoid the `goto`? :)

Comment (abtink) [reply src/core/utils/parse_cmdline.cpp]: Changed to use a boolean flag `skipFirstDigit`.

**Review by jwhui — APPROVED** [review]

**Pull request merged**